### ISSUE TITLE
Suggestion: Permit an implementing class to ignore private methods of the implementee class

### ISSUE THREAD

**Issue by NoelAbrahams** [Microsoft/TypeScript#471]
Hi, we have a requirement to create mocks of real classes defined in code. The exact requirements are

* A specific mock should define the public API of the real class.
* If the public API of the real class changes in any way then the compiler should detect the corresponding breakage in the mock class.

A possible solution might be as follows: 
```TypeScript
declare var fileWriter: any; 

// Real class
class Foo {
	
	public writeToFile(){
		fileWriter.writeToFile('');
	}
}

// Mock
class MockFoo implements Foo {
	 
	 public writeToFile(){
		 // do nothing
	 }
}
``` 
This appears to solve the problem because changing `Foo.writeToFile` will trigger a compilation error along the lines of "Class MockFoo declares interface Foo but does not implement it..."

The problem with this approach is that `class Foo` is not permitted to have any private methods or fields. If we were to add a private method `foo()` to `class Foo` then it's no longer possible for `MockFoo` to implement `Foo` because the compiler doesn't permit it.

I suggest that when a class implements another class the implementing class be allowed to ignore the private fields and methods (both static and instance) of the implementee.

(I am aware that there are workarounds, such as declaring an interface that both `Foo` and `MockFoo` implement, but that introduces an unnecessary maintenance overhead.)

**Comment by aciccarello** [comment]
I am still trying to understand what the language of `implementingClass implements baseClass` means and am not following why private attributes need to be in the implementing class.

Can someone reiterate why you can't just "ignore private variables when implementing interfaces".

I agree with @NoelAbrahams that I look at interfaces as a public API definition.
>To me the idea of being able to use the public API of a class as an interfaces is quite appealing; and probably has many interesting use-cases in the real world.

**Comment by DanielRosenwasser** [comment]
> Can someone reiterate why you can't just "ignore private variables when implementing interfaces".

Even though it isn't distinguishable whether a type has privates from *outside* the class, an instance of `Foo` is allowed to access the privates of any other `Foo`. For this reason, it needs to be the case that when implementing `Foo`, the privates need to be there or else a method in `Foo` using the private of another `Foo` will fail.

See @RyanCavanaugh's example in the above https://github.com/Microsoft/TypeScript/issues/471#issuecomment-52807450.

**Comment by aciccarello** [comment]
@DanielRosenwasser Okay thank you. That clarifies it for me. I saw that example but was confused because I was under the impression that `Foo` could not access the private variables of another instance.

**Comment by benliddicott** [comment]
I've opened a related issue #3854 which would solve the originators motivating problem.

**Comment by benliddicott** [comment]
see alsp #4278

**Comment by benliddicott** [comment]
This is NeedsProposal. Here's a proposal which would allow the workaround posted to #3854  and #4278  work for `private` members as well as `protected` members.

### Proposal: Allow an extending class to modify the protection of a private member.

Currently an extending class can modify a `protected` member to be `public`:

```ts

class X{
    protected m1: string;
}
class Y extends X {
    public m1: string;
}

```
The proposal is to allow the same syntax to expose a `private` member as `protected` or `public`.

**Comment by RyanCavanaugh** [comment]
These are reasonable use cases, but the problem is that we don't want `class X implements Y { ... }` to produce an `X` that is not assignable to `Y` (as would be the case if this worked as proposed).

If we get a general notion of type operators (many issues wanting this), a simple operator e.g. `publicShapeOf` could solve these use cases.

**Comment by Elephant-Vessel** [comment]
I'm trying to understand the motivations regarding the language design here. Again I wish to reduce brain pain, hope you don't mind.

I _do_ understand why privates currently need to be exposed, as the scope of the privates are bound to the class and not the instance. I _don't_ understand, considering the consequences, why the scope of privates are bound to the class and not to the instance.

Is there any significant value in that a Foo can manipulate the implementation details of any Foo? Something that makes up for the non-intuitiveness and the disturbed ability of the language to easily provide abstractional reusability that comes from mixing in privates with types/interfaces?

I would like to be able to explain to my colleagues what makes this strange deal with exposed privates worth it after all.

**Comment by Elephant-Vessel** [comment]
> > Why not simply allow implementing private members of class ?

> Yes, this would be an improvement on the current situation, because as it stands the idea of "implementing another class" is not useful.

Although that would further complicate this whole concept of privates and what it actually means.

Same thing with letting subclasses upgrade private and protected fields. Wasn't even aware of that subclasses already can update protected fields, feels like a complete violation of why we have those modifiers in the first place. Feels like this can get messy.

**Comment by RyanCavanaugh** [comment]
@Elephant-Vessel 
> I do understand why privates currently need to be exposed, as the scope of the privates are bound to the class and not the instance. I don't understand, considering the consequences, why the scope of privates are bound to the class and not to the instance.

Certainly TypeScript is not the first language deciding to scope `private` to the type rather than the instance. The trade-offs are fairly obvious IMO. Probably it's going to be easier to find discussion of the merits of this when applied to Java, C#, C++, etc. rather than us going over that again here.

**Comment by Elephant-Vessel** [comment]
@RyanCavanaugh 
> Certainly TypeScript is not the first language deciding to scope private to the type rather than the instance. The trade-offs are fairly obvious IMO. Probably it's going to be easier to find discussion of the merits of this when applied to Java, C#, C++, etc. rather than us going over that again here.

Thank you for your response Ryan.

Sure, but TypeScript differs from those languages as it's structurally typed. And that means that we suddenly have this problem with privates having to be propagated in the type system, a problem that these other languages do not experience.

I personally believe that solid abstractural integrity is significantly more important than allowing a Foo to mess around with the internals of other Foos. But if the reason for this is that the other languages do it, then I at least know what to say to my colleagues. Brain still not happy, but less pain. Thanks.

**Comment by kitsonk** [comment]
> I personally believe that solid abstractural integrity is significantly more important than allowing a Foo to mess around with the internals of other Foos.

I don't think a class having access to/visibility of other instances breaks encapsulation at all.  The proposal for ES private properties also includes this class level visibility.  It is simply a different form of logical *state* which the class has access to.  I like to think of this private state as a class visible WeakMap, and in fact for @dojo this is actually what we do with our private state, versus using `private` properties, which ensures at run-time private is really private, as we don't leak our WeakMaps outside of the module the class is defined in.

The main use case is that the class having access to the private properties allows for static methods to potentially operate on instances without needing to break this encapsulation.

**Comment by Elephant-Vessel** [comment]
@kitsonk 
Sure, in regards of design constraints the current policy is more a nuisance in regards of reusability than breaking abstractions*. But I think it makes the conceptual model of the language unnecessary convoluted  as we now have this unintuitive behavior and the mixup of public contracts with internal implementation within abstractions. How easy wouldn't it have been if we had instance privates instead of class privates, none of this would have been an issue, and the conceptual model would be less complex. May it be that those static methods would have to be instance methods instead.

* And then we have this stuff about promoting protected members, which I'm really curious about why we got in the first place, and the suggestion about doing the same thing with private members. Talk about breaking encapsulation. For a language whose selling point is to be able to scale, this is beyond me.

**Comment by ricklove** [comment]
I like the idea of `publicShapeOf` where can I track that?

> If we get a general notion of type operators (many issues wanting this), a simple operator e.g. `publicShapeOf` could solve these use cases.
From: https://github.com/Microsoft/TypeScript/issues/471#issuecomment-196583994

**Comment by AlexGalays** [comment]
If this problem is going to remain, you should at least prevent class from "implementing" other classes, imo.

**Comment by DanielRosenwasser** [comment]
Since mapped object types strip out `private`/`protected` fields (though they also strip out call/construct signatures), @wycats had the idea of writing the following to achieve this:

```ts
type Interface<T> = {
    [P in keyof T]: T[P]
}

class C {
    private foo(): string;

    bar(): number;
}

// Look ma, no 'foo'!
class D implements Interface<C> {
    bar() {
        // ...
    }
}
```

I've also called that mapped object type `PropsOf`, and you could call it `PublicOf` if you so desired.
